add ui element markdown-parser and markdown support on metadata…

cdebarros opened the pull request:
Hi, first of all happy new year to everyone seeing this. Here is my first contribution to the project.

**This pull request makes the following changes :**
* Fixes isssue #733 

I used `marked` to parse markdown content to html and then i defined some basic style for the following elements as suggested by @jahow  :

- h1 to h6
- ul, ol and li
- p
- a
- blockquote
- pre - code
- hr
- table
- img

I tried to use tailwind base colors to make it close to the default style. 
I'm not sure how should i render links, blockquotes or table. I'll wait for your feedbacks to make some changes if needed.

fgravin reviewed (commented):
Thanks @cdebarros for this first PR.

I need to wrap up the approach a bit with @jahow but as a first review, I would just point out that a component seems not appropriate for such a process.

I would have more seen a Pipe and/or a directive.

Also, I wonder if the css could be encapsulated somehow instead of being spread via the `encapsulation: ViewEncapsulation.None` option.

jahow commented:
> I need to wrap up the approach a bit with @jahow but as a first review, I would just point out that a component seems not appropriate for such a process.
> 
> I would have more seen a Pipe and/or a directive.

What would be an argument for using a pipe or directive instead of a component? I'm still unsure which solution is the best personally. Using a component has the advantage that we can use an encapsulated style inside it.

jahow commented on libs/ui/elements/src/lib/markdown-parser/markdown-parser.component.css:
```diff
@@ -0,0 +1,271 @@
+  line-height: 1.25;
+}
+
+.markdown-body h1 {
+  margin: 0.67em 0;
+  font-weight: 600;
+  padding-bottom: 0.3em;
+  font-size: 2em;
+  color: var(--color-primary);
+}
```
The title styles currently are inconsistent with other titles, see:
![image](https://github.com/geonetwork/geonetwork-ui/assets/10629150/8196d772-dbd6-4550-8202-d22563f4773c)

I would suggest using `@apply text-title font-title` for all `hx` classes above

jahow commented on libs/ui/elements/src/lib/markdown-parser/markdown-parser.component.ts:
```diff
@@ -0,0 +1,16 @@
+  styleUrls: ['./markdown-parser.component.css'],
+  encapsulation: ViewEncapsulation.None,
+})
+export class MarkdownParserComponent {
```
We usually try to add presentation components to storybook, do you think you could do the same with this one?
I suggest you take inspiration from other components such as this one: https://github.com/geonetwork/geonetwork-ui/blob/main/libs/ui/elements/src/lib/metadata-info/linkify.directive.stories.ts

The goal is to let people easily "try" the component and see its behaviour with varying inputs, and also making sure that it doesn't break after modifications.

jahow commented on libs/ui/elements/src/lib/markdown-parser/markdown-parser.component.ts:
```diff
@@ -0,0 +1,16 @@
+  selector: 'gn-ui-markdown-parser',
+  templateUrl: './markdown-parser.component.html',
+  styleUrls: ['./markdown-parser.component.css'],
+  encapsulation: ViewEncapsulation.None,
```
you can get rid of this and leave the default view encapsulation by adding `:host /deep/` before all classes in the CSS file; this is a bit more verbose but IMO essential since disabling view encapsulation is risky in the long run

jahow reviewed (commented):
Thanks for this contribution! I think this looks really good. I added a few comments, specifically to avoid disabling the view encapsulation; once addressed I think this can be merged. Please let me know if you need any more information!

@fgravin I've thought about it again and using a component seems right to me; please let us know if you have arguments against it

jahow commented:
Note: Snyk workflow always fails for forks; please disregard that

cdebarros commented:
Thanks for your reviews ! 
i will try to make these changes when i have time, probably before the end of this week.

cdebarros commented:
I just updated my contribution. Added css encapsulation and storybook for the component.

On the default storybook content, i explain which markdown elements are supported for now.

jahow commented on libs/ui/elements/src/lib/markdown-parser/markdown-parser.component.spec.ts:
```diff
@@ -0,0 +1,21 @@
+
+  it('should create', () => {
+    expect(component).toBeTruthy()
+  })
```
Oh sorry, I didn't realize that in the first review but we we need at least one test here, showing that a markdown input will be correctly transformed to HTML; this doesn't have to be exhaustive as we're not testing the underlying library, but simply making sure that the test is processed.

Could you please add that?

cdebarros replied:
Oh, i totally forgot about that.

I created 3 tests described like this :
- should render markdown as HTML
- should render HTML as HTML
- should render text as HTML 

Is it too much ? Maybe the first test is enough.

jahow replied:
That does look like a perfect amount of tests :slightly_smiling_face:

jahow commented on libs/ui/elements/src/lib/markdown-parser/markdown-parser.component.ts:
```diff
@@ -0,0 +1,15 @@
+@Component({
+  selector: 'gn-ui-markdown-parser',
+  templateUrl: './markdown-parser.component.html',
+  styleUrls: ['./markdown-parser.component.css'],
```
I also forgot to point out that we should always use `changeDetection: ChangeDetectionStrategy.OnPush,` to make sure that the change detection does not run too often

jahow reviewed (commented):
Thanks for the adaptations, I think this looks better now. There are choices in the styling of the output that might not fit for every occasion and theme, but for now we can keep it like that and see how it behaves in the "wild".

I realized there was no unit test, I missed that on the first review. Could you please add one that ensures that the input is processed into the div inner HTML?

Also the branch needs to be rebased; there will likely be a conflict because we added `gnUiLinkify` on the abstract in the meantime; this directive can be removed from the abstract since `marked` does the same thing :)

Thank you for your continued efforts

cdebarros commented:
I added tests and the changeDetectionStrategy to the component. Also i rebased the branch.
Let me know if it needs other changes.

jahow reviewed (approved):
Thanks for this great contribution! There is a failure on the build attempts in the CI, to fix it you can bump the CSS budgets for all apps like so:
```diff
diff --git a/apps/datafeeder/project.json b/apps/datafeeder/project.json
index 67998c55..b22bc758 100644
--- a/apps/datafeeder/project.json
+++ b/apps/datafeeder/project.json
@@ -40,8 +40,8 @@
             },
             {
               "type": "anyComponentStyle",
-              "maximumWarning": "2kb",
-              "maximumError": "4kb"
+              "maximumWarning": "10kb",
+              "maximumError": "20kb"
             }
           ],
           "fileReplacements": [
```

cdebarros commented:
I updated it for the 3 projects that have failed, it should be fine now.

jahow commented:
Don't worry about the CSS warnings, I'll fix them in another PR.

Thank you again for this awesome contribution, congratulations on it and we're looking forward to more of this!! :tada:
